File: src/main/java/com/aws/proserve/ggcommons/messaging/MessageBuilder.java
package com.aws.proserve.ggcommons.messaging;

import com.aws.proserve.ggcommons.config.ConfigManager;
import com.google.gson.Gson;
import com.google.gson.JsonSyntaxException;

/**
 * Builder for creating Message instances with fluent API.
 */
public class MessageBuilder {
    private String name;
    private String version;
    private String correlationId;
    private Object payload;
    private ConfigManager configManager;
    
    private MessageBuilder(String name, String version) {
        this.name = name;
        this.version = version;
    }
    
    public static MessageBuilder create(String name, String version) {
        return new MessageBuilder(name, version);
    }
    

    
    public MessageBuilder withCorrelationId(String correlationId) {
        this.correlationId = correlationId;
        return this;
    }
    
    public MessageBuilder withPayload(Object payload) {
        this.payload = payload;
        return this;
    }
    
    public MessageBuilder withConfig(ConfigManager configManager) {
        this.configManager = configManager;
        return this;
    }
    
    public Message build() {
        if (configManager == null) {
            throw new IllegalStateException("ConfigManager is required - call withConfig()");
        }
        
        Message message = new Message();
        message.header = new MessageHeader(name, version, correlationId);
        message.tags = MessageTags.fromConfig(configManager);
        
        if (payload instanceof String) {
            String payloadStr = (String) payload;
            try {
                Gson gson = new Gson();
                message.body = gson.fromJson(payloadStr, Object.class);
            } catch (JsonSyntaxException e) {
                message.body = payloadStr;
            }
        } else {
            message.body = payload;
        }
        
        return message;
    }
}

File: ggcommons/messaging/message_builder.py
"""
Builder for creating Message instances with fluent API.
"""
import json
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ggcommons.config.manager.config_manager import ConfigManager


class MessageBuilder:
    def __init__(self, name: str, version: str):
        self.name = name
        self.version = version
        self.correlation_id = None
        self.uuid = None
        self.timestamp = None
        self.reply_to = None
        self.payload = None
        self.config_service = None
        self.tags = None

    @staticmethod
    def create(name: str, version: str) -> 'MessageBuilder':
        return MessageBuilder(name, version)

    @staticmethod
    def from_object(msg_contents) -> 'MessageBuilder':
        if isinstance(msg_contents, dict) and "header" in msg_contents:
            header = msg_contents["header"]
            name = header.get("name", "unknown")
            version = header.get("version", "1.0")
            
            builder = MessageBuilder(name, version)
            
            if "correlation_id" in header:
                builder.with_correlation_id(header["correlation_id"])
            if "uuid" in header:
                builder.with_uuid(header["uuid"])
            if "timestamp" in header:
                builder.with_timestamp(header["timestamp"])
            if "reply_to" in header:
                builder.with_reply_to(header["reply_to"])
            if "body" in msg_contents:
                builder.with_payload(msg_contents["body"])
            if "tags" in msg_contents:
                builder.with_tags(msg_contents["tags"])
            
            return builder
        else:
            # Raw message without header structure
            builder = MessageBuilder("raw", "1.0")
            builder.with_payload(msg_contents)
            return builder

    def with_correlation_id(self, correlation_id: str) -> 'MessageBuilder':
        self.correlation_id = correlation_id
        return self

    def with_payload(self, payload) -> 'MessageBuilder':
        self.payload = payload
        return self

    def with_config(self, config_service: 'ConfigManager') -> 'MessageBuilder':
        self.config_service = config_service
        return self

    def with_tags(self, tags: dict) -> 'MessageBuilder':
        self.tags = tags
        return self

    def with_uuid(self, uuid: str) -> 'MessageBuilder':
        self.uuid = uuid
        return self

    def with_timestamp(self, timestamp: str) -> 'MessageBuilder':
        self.timestamp = timestamp
        return self

    def with_reply_to(self, reply_to: str) -> 'MessageBuilder':
        self.reply_to = reply_to
        return self

    def build(self):
        if self.config_service is None and self.tags is None:
            raise ValueError("Configuration service is required - call with_config()")

        from ggcommons.messaging.message import Message, MessageHeader, MessageTags

        message = Message()
        message.header = MessageHeader(self.name, self.version, self.correlation_id, self.timestamp, self.uuid, self.reply_to)
        
        if self.tags is not None:
            message.tags = MessageTags.from_dict(self.tags)
        elif self.config_service is not None:
            message.tags = MessageTags.from_config(self.config_service)
        
        if isinstance(self.payload, str):
            try:
                message.body = json.loads(self.payload)
            except json.JSONDecodeError:
                message.body = self.payload
        else:
            message.body = self.payload

        return message